### PR TITLE
fix halo exchange bug [closes #1689]

### LINKED ISSUE #1689: [LFRic] Incorrect stencil depth list for discontinuous stencil readers followed by discontinuous writers
When we have a kernel that has discontinuous stencil readers is followed by a kernel that has discontinuous writers in a single invoke, PSyclone generates incorrect code for stencil depth list string. E.g. as reported from [LFRic/branches/dev/ianboutle/r34895_bl_exp_split@35124](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/branches/dev/ianboutle/r34895_bl_exp_split?rev=35124) ([LFRic ticket 3205](https://code.metoffice.gov.uk/trac/lfric/ticket/3205)):
```
algorithm/bl_exp_alg_mod_psy.f90(493): error #6642: There are null arguments in an intrinsic function reference.   [MAX]
      IF (u_in_w3_proxy%is_dirty(depth=max(stencil_depth,))) THEN
---------------------------------------^
algorithm/bl_exp_alg_mod_psy.f90(494): error #6642: There are null arguments in an intrinsic function reference.   [MAX]
        CALL u_in_w3_proxy%halo_exchange(depth=max(stencil_depth,))
-----------------------------------------------^
algorithm/bl_exp_alg_mod_psy.f90(497): error #6642: There are null arguments in an intrinsic function reference.   [MAX]
      IF (v_in_w3_proxy%is_dirty(depth=max(stencil_depth,))) THEN
---------------------------------------^
algorithm/bl_exp_alg_mod_psy.f90(498): error #6642: There are null arguments in an intrinsic function reference.   [MAX]
        CALL v_in_w3_proxy%halo_exchange(depth=max(stencil_depth,))
-----------------------------------------------^
algorithm/bl_exp_alg_mod_psy.f90(501): error #6642: There are null arguments in an intrinsic function reference.   [MAX]
      IF (tile_fraction_proxy%is_dirty(depth=max(stencil_depth,))) THEN
---------------------------------------------^
algorithm/bl_exp_alg_mod_psy.f90(502): error #6642: There are null arguments in an intrinsic function reference.   [MAX]
        CALL tile_fraction_proxy%halo_exchange(depth=max(stencil_depth,))
```

Closer inspection into the `dynamo0p3.py` function [`def _compute_halo_depth`](https://github.com/stfc/PSyclone/blob/ffd115beb3990679bff78e390f9909353ba073b0/src/psyclone/dynamo0p3.py#L5950) reveals that the problem lies in this check
```
        if len(depth_info_list) == 1:
            return str(depth_info_list[0])
```
because call to [`_compute_halo_read_depth_info`](https://github.com/stfc/PSyclone/blob/ffd115beb3990679bff78e390f9909353ba073b0/src/psyclone/dynamo0p3.py#L5961) may return an **empty string** which is **still counted as a list element**.

A print-out of `depth_info_list` and `depth_str_list` for a combination of two test-file kernels
```
 call invoke(                                                       &
       testkern_different_any_dscnt_space_stencil_type(f1,          &
                                                       f2, extent,  &
                                                       f4, extent), &
       testkern_w3_type(a, f1, f2, f3, f4)                          &
       )
``` 
returns something like 
```
depth_info_list:  [<psyclone.dynamo0p3.HaloDepth object at 0x7fe4d23fc710>, <psyclone.dynamo0p3.HaloDepth object at 0x7fe4d23fcf60>]
depth_str_list:  ['extent', '1']
depth_info_list:  [<psyclone.dynamo0p3.HaloDepth object at 0x7fe4d2272940>, <psyclone.dynamo0p3.HaloDepth object at 0x7fe4d2272400>]
depth_str_list:  ['extent', '']
```

Interestingly, this problem doesn't appear when
- The kernel with discontinuous stencil access is placed after a kernel with discontinuous writers in a single invoke;
- The kernel with discontinuous stencil access and kernel with discontinuous writers are in different invokes;
-  One or both kernels have continuous stencil access or/and readers.

A simple fix is to first create `depth_str_list` from `depth_info_list`, filter it for empty entries and then check the length of `depth_str_list`.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for 1st review. Probably one for @arporter or @TeranIvy

**Comment by TeranIvy** [comment]
I'm happy to take this so I can check it in LFRic.

**Comment by TeranIvy** [comment]
Closed #1690 as this PR supersedes it.

File: src/psyclone/dynamo0p3.py [hunk 1]
```diff
@@ -5886,13 +5886,14 @@ def _create_depth_list(halo_info_list, sym_table):
...
                 break
         if not match:
-            # no matches were found with existing entries so
-            # create a new one
-            depth_info = HaloDepth(sym_table)
-            depth_info.set_by_value(max_depth=False, var_depth=var_depth,
-                                    literal_depth=literal_depth,
-                                    annexed_only=False, max_depth_m1=False)
-            depth_info_list.append(depth_info)
+            # no matches were found with existing entries so create a
```
Comment (TeranIvy) [inline src/psyclone/dynamo0p3.py:5889]: I would suggest starting the comment with upper case.

File: src/psyclone/dynamo0p3.py [hunk 2]
```diff
@@ -5886,13 +5886,14 @@ def _create_depth_list(halo_info_list, sym_table):
...
+            # new one (unless no var depth and literal_depth is 0).
```
Comment (TeranIvy) [inline src/psyclone/dynamo0p3.py:5890]: I would suggest "unless there is no 'var_depth' and 'literal_depth' is 0)."

File: src/psyclone/tests/dynamo0p3_haloex_test.py [hunk 3]
```diff
@@ -36,20 +36,20 @@
...
-from psyclone.parse.algorithm import parse
-from psyclone.psyGen import PSyFactory, GenerationError
-from psyclone.dynamo0p3 import DynLoop, DynHaloExchange
-from psyclone.transformations import Dynamo0p3RedundantComputationTrans, \
-    Dynamo0p3AsyncHaloExchangeTrans
 from psyclone.configuration import Config
-from psyclone.errors import InternalError
 from psyclone.core.access_info import AccessType
-
+from psyclone.dynamo0p3 import (
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_haloex_test.py:44]: Thanks for cleaning up the imports 😄

File: src/psyclone/tests/dynamo0p3_haloex_test.py [hunk 4]
```diff
@@ -589,3 +589,51 @@ def test_gh_readinc(tmpdir):
     assert f1_hex._compute_halo_depth() == '1'
 
     assert LFRicBuild(tmpdir).code_compiles(psy)
+
+
+def test_stencil_then_w3_read(tmpdir):
+    '''Test that a stencil access to a discontinuous field followed by a
+    read (or readwrite) in a subsequent kernel to the same field
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_haloex_test.py:596]: I would suggest "'read' (or 'readwrite') access".

File: src/psyclone/tests/dynamo0p3_haloex_test.py [hunk 5]
```diff
@@ -589,3 +589,51 @@ def test_gh_readinc(tmpdir):
...
+
+    '''
+    # Check that an instance of the HaloDepth class returns the
+    # expected results for this case.
+    halo_depth = HaloDepth(None)
+    assert str(halo_depth) == "0"
+    halo_depth2 = HaloDepth(None)
+    halo_depth2._var_depth = "extent"
+    assert str(halo_depth2) == "extent"
+    # Quick check when we have both depth and literal depth>0
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_haloex_test.py:610]: I would suggest "depth > 0" (blank spaces).

File: src/psyclone/tests/dynamo0p3_haloex_test.py [hunk 6]
```diff
@@ -589,3 +589,51 @@ def test_gh_readinc(tmpdir):
...
+    halo_depth2.literal_depth = 1
+    assert str(halo_depth2) == "extent+1"
+    # Go back to original 0 depth case
+    halo_depth2.literal_depth = 0
+
+    # Check that _create_depth_list removes depths that are 0 from its
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_haloex_test.py:616]: I would suggest single quotations around "'_create_depth_list'".

File: src/psyclone/tests/dynamo0p3_haloex_test.py [hunk 7]
```diff
@@ -589,3 +589,51 @@ def test_gh_readinc(tmpdir):
...
+    # return list. It takes two entries as input and returns one.
+    result = _create_depth_list([halo_depth, halo_depth2], None)
+    assert isinstance(result, list)
+    assert len(result) == 1
+    assert isinstance(result[0], HaloDepth)
+    assert str(result[0]) == "extent"
+
+    # Check it all works in practice (functional test)
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_haloex_test.py:624]: I would suggest "Check that it all works ..."

File: src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90 [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. W. Ford, STFC Daresbury Lab
+
+program stencil_then_read
+
+  ! Test that a discontinuous stencil read (which happens to be
```
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90:38]: I would suggest single quotations around "'read'".

File: src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90 [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  ! any_discontinuous but does not matter) followed by a discontinuous
```
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90:39]: I would suggest single quotations around "'any_discontinuous'".

File: src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90 [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  ! read (which happens to be readwrite but does not matter) results
```
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90:40]: I would suggest single quotations around "'readwrite'".

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Thank you very much for the fix, it is much better solution than my hacky attempt!
I have a few very minor nitpicky comments (inline). I would also request the branch be updated with the latest master.
- Tests with compilation pass.
- Examples run.
- Changes are `pycodestyle` and `pylint` clean.
- Changes are covered by tests.
- There is no change to the documentation.

**Comment by TeranIvy** [comment]
@rupertford, sorry for taking so long to review the PR! I have some very minor remarks and after that we are ready to go 😄

File: src/psyclone/dynamo0p3.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/dynamo0p3.py:5889]: Updated this and all other comments in this function.

File: src/psyclone/dynamo0p3.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/dynamo0p3.py:5890]: Updated.

File: src/psyclone/tests/dynamo0p3_haloex_test.py (hunk 4 above)
Comment (rupertford) [reply src/psyclone/tests/dynamo0p3_haloex_test.py:596]: OK.

File: src/psyclone/tests/dynamo0p3_haloex_test.py (hunk 5 above)
Comment (rupertford) [reply src/psyclone/tests/dynamo0p3_haloex_test.py:610]: Added spaces.

File: src/psyclone/tests/dynamo0p3_haloex_test.py (hunk 6 above)
Comment (rupertford) [reply src/psyclone/tests/dynamo0p3_haloex_test.py:616]: quotationified.

File: src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90 (hunk 8 above)
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90:38]: Added.

File: src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90 (hunk 9 above)
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90:39]: Added. Also added _space.

File: src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90 (hunk 10 above)
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/14.16_disc_stencil_then_read.f90:40]: Done.

**Comment by rupertford** [comment]
Thanks @TeranIvy. This is ready for its next review.

**Review by TeranIvy — APPROVED** [review]
All well now. Proceeding to merge to master.

**Pull request merged**